flash/partition_info/partition_info.c

```diff
@@ -0,0 +1,162 @@
+    uint32_t unpartitioned_space_first_sector;
+    uint32_t unpartitioned_space_last_sector;
+    uint32_t flags_and_permissions;
+    int current_partition;
+    size_t pos;
+    int status;
+} pico_partition_table_t;
+
+/*
+ * Stores information on each partition
+ */
+typedef struct {
+    uint32_t first_sector;
+    uint32_t last_sector;
+    uint32_t flags_and_permissions;
+    uint64_t partition_id;
+    char name[PARTITION_NAME_MAX + 1];  // name length is indicated by 7 bits
+} pico_partition_t;
+
+
+/*
+ * Read the partition table information.
+ *
+ * See the RP2350 datasheet 5.1.2, 5.4.8.16 for flags and structures that can be specified.
+ */
+int open_partition_table(pico_partition_table_t *pt) {
+    // Reads fixed size fields
+    uint32_t flags = PT_INFO_PT_INFO | PT_INFO_PARTITION_LOCATION_AND_FLAGS | PT_INFO_PARTITION_ID;
+    int rc = rom_get_partition_table_info(pt->table, sizeof(pt->table), flags);
+    if (rc < 0) {
+        pt->partition_count = 0;
+        pt->status = rc;
+        return rc;
+    }
+
+    size_t pos = 0;
+    pt->fields = pt->table[pos++];
+    assert(pt->fields == flags);
+    pt->partition_count = pt->table[pos++] & 0x000000FF;
+    uint32_t location = pt->table[pos++];
+    pt->unpartitioned_space_first_sector = PART_LOC_FIRST(location);
+    pt->unpartitioned_space_last_sector = PART_LOC_LAST(location);
+    pt->flags_and_permissions = pt->table[pos++];
+    pt->current_partition = 0;
+    pt->pos = pos;
+    pt->status = 0;
+
+    return 0;
+}
+
+/*
+ * Extract each partition information
+ */
+bool read_next_partition(pico_partition_table_t *pt, pico_partition_t *p) {
+    if (pt->current_partition >= pt->partition_count) {
+        return false;
+    }
+
+    size_t pos = pt->pos;
+    uint32_t location = pt->table[pos++];
+    p->first_sector = PART_LOC_FIRST(location);
+    p->last_sector = PART_LOC_LAST(location);
+    p->flags_and_permissions = pt->table[pos++];
+
+    if (p->flags_and_permissions & PICOBIN_PARTITION_FLAGS_HAS_ID_BITS) {
+        uint32_t id_low  = pt->table[pos++];
+        uint32_t id_high = pt->table[pos++];
+        p->partition_id = ((uint64_t)id_high << 32) | id_low;
+    } else {
+        p->partition_id = 0;
+    }
+    pt->pos = pos;
+
+    if (p->flags_and_permissions & PICOBIN_PARTITION_FLAGS_HAS_NAME_BITS) {
+        // Read variable length fields
+        uint32_t name_buf[(PARTITION_NAME_MAX + 1) / sizeof(uint32_t)] = {0};
+        uint32_t flags = PT_INFO_SINGLE_PARTITION | PT_INFO_PARTITION_NAME;
+        int rc = rom_get_partition_table_info(name_buf, sizeof(name_buf),
+                                              (pt->current_partition << 24 | flags));
+        if (rc < 0) {
+            pt->status = rc;
+            return false;
+        }
+        uint32_t __attribute__((unused)) fields = name_buf[0];
+        assert(fields == flags);
+        uint8_t *name_buf_u8 = (uint8_t *)&name_buf[1];
+        uint8_t name_length = *name_buf_u8++;
```

Comment:
Datasheet says "Byte 0 : 7 bit length of the name (LEN); top bit reserved" so perhaps there ought to be a ` & 0x7F` here?